added the TileApi to control real colorful lifx tiles

mabels opened the pull request:
I need to cleanup this PR some sanity checks are missing and 
i was to lazy to build the tests.
I promise that I complete this missing stuff in the next weeks.
This PR is just an reminder to me that i had to completed this!

mabels committed:
added the TileApi to control real colorful lifx tiles

ristomatti commented:
@mabels Based on a quick skim, your PR looks very promising! When finalising the work, please remember to:
- update `README.md`
- follow `.eslintrc` 

Also note that you can use arrow functions even though it's not used much if at all on the existing codebase (the ES2015 rewrite was done quite literally overnight). At least I would prefer arrow functions especially when passing an inline function as a parameter. I can't speak for @Sawtaytoes though who set up this fork though.

mabels commented:
Hi,

just completed this including tests. There had been 2 other pull request which should applied first.

https://github.com/node-lifx/lifx-lan-client/pull/16
https://github.com/node-lifx/lifx-lan-client/pull/17

Thx in advance

ristomatti commented:
Massive undertaking. I'll try to review and test this as soon as possible. Sorry if it might take a couple of days! I do have a Tile set to test though.

ristomatti commented:
@mabels I'm sorry I haven't still had time to review and test this. Just letting you know I haven't forgot this! I just noticed a PR submitted to **node-lifx-lan** a couple of days ago and it'd be cool to have this merged before that. :joy: The problem is though that I'm quite busy today and away from home the rest of the weekend. I'll try to find time today but it doesn't look too good.

ristomatti commented:
@mabels BTW just noticed there seems to be a conflict due to merging your previous PR #17.

ristomatti commented:
My plans for today/tomorrow got cancelled. I'll at least start reviewing/testing now.

ristomatti commented:
@mabels All you need to do to fix the conflict is to drop the already merged commits 3de27da and 
1cca93a and then rebase to upstream master.

ristomatti commented:
It's actually not a good idea to review the code at it's current state as it contains the already merged changes. @mabels could you update your branch and force push the changes? I'll still test now if I get it to work with my Tiles.

ristomatti commented:
Sorry a somewhat off-topic note:

> Also note that you can use arrow functions even though it's not used much if at all on the existing codebase (the ES2015 rewrite was done quite literally overnight). At least I would prefer arrow functions especially when passing an inline function as a parameter. I can't speak for @Sawtaytoes though who set up this fork though.

Quoting myself - now when I started to review the changes @mabels has done, I realize there's still two open PR's from me that's aren't included in this repo. When commenting about the arrow functions, I had an odd feeling like I would have done "something" to that already: https://github.com/MariusRumpf/node-lifx/pull/62. This is the *rewrite* I was talking about.

@Sawtaytoes It looks like I forgot PR 62 completely. Would it make sense for me to try to rebase it and do a PR? Or do you prefer working without the use of classes. I've personally been leaning towards more functinal programming approach (mainly at work) but I think using classes would make sense given how the library is written. There'd be a lot less clutter.

ristomatti commented on src/lifx/light.js:
```diff
@@ -71,9 +71,9 @@ Light.prototype.color = function(hue, saturation, brightness, kelvin, duration,
-  brightness = Math.round(brightness / constants.HSBK_MAXIMUM_BRIGHTNESS * 65535);
+  hue = Math.round((hue / constants.HSBK_MAXIMUM_HUE) * 65535);
+  saturation = Math.round((saturation / constants.HSBK_MAXIMUM_SATURATION) * 65535);
+  brightness = Math.round((brightness / constants.HSBK_MAXIMUM_BRIGHTNESS) * 65535);
```
Revert these unrelated changes. It'll be really confusing to see `git blame` say you've last edited the lines but the commit says something about "TileApi".

This would improve the readability a bit I agree, but as this `* 655535` part is repeating so much it would make more sense to refactor it to be done in a function anyway. (But let's still keep refactoring out from this PR).

mabels replied:
I revert this changes, but it drives me crazy.... infact nobody normal knows the order of execution of the operators * /.

ristomatti commented on src/lifx/light.js:
```diff
@@ -134,7 +134,7 @@ Light.prototype.colorRgbHex = function(hexString, duration, callback) {
   validate.irBrightness(brightness, 'light setMaxIR method');
 
-  brightness = Math.round(brightness / constants.IR_MAXIMUM_BRIGHTNESS * 65535);
+  brightness = Math.round((brightness / constants.IR_MAXIMUM_BRIGHTNESS) * 65535);
```
Revert these unrelated changes.

mabels replied:
I revert this changes, but it drives me crazy.... infact nobody normal knows the order of execution of the operators * /.

ristomatti commented on src/lifx/light.js:
```diff
@@ -510,9 +510,9 @@ Light.prototype.colorZones = function(startIndex, endIndex, hue, saturation, bri
-  brightness = Math.round(brightness / constants.HSBK_MAXIMUM_BRIGHTNESS * 65535);
+  hue = Math.round((hue / constants.HSBK_MAXIMUM_HUE) * 65535);
+  saturation = Math.round((saturation / constants.HSBK_MAXIMUM_SATURATION) * 65535);
+  brightness = Math.round((brightness / constants.HSBK_MAXIMUM_BRIGHTNESS) * 65535);
```
Revert these unrelated changes.

mabels replied:
I revert this changes, but it drives me crazy.... infact nobody normal knows the order of execution of the operators * /.

ristomatti commented on src/lifx/packets/stateDeviceChain.js:
```diff
@@ -0,0 +1,79 @@
+  size:
+    1 +
+    16 * (2 + 2 + 2 + 2 + 4 + 4 + 1 + 1 + 1 + 4 + 4 + 4 + 8 + 8 + 2 + 2 + 4) +
+    1,
```
The `size` is quite confusing like this. Just use the end result value like on the other packets.

Sawtaytoes replied:
I think I get what you're saying @ristomatti. Since those are the UDP groupings from LIFX's docs, I'd instead recommend a comment like this:

```js
size: X // https://docs.lifx.com/some-udp-article
```

Or even:

```js
size: X // This is a grouping of 2, 2, 2, 2, 4, 4, 1, ... etc groupings from LIFX's UDP protocol.
```

ristomatti replied:
I link to the relevant section of the LAN protocol specification would be sufficient. We don't want to duplicate the documentation here as it occasionally also changes.

Some time ago LIFX has published a machine readable spec document that should make it possible to generate parts of the code. That might be another angle to look into if we'd like to get away with these relatively cryptic packet files. The naming follows the names used in the spec document though so it's quite easy to figure out each part.

mabels replied:
I simply changed it to the value.
To put such things in the comments is not helpful.
So we lose the source of this arbitrated value, i can live with this.

ristomatti commented on src/lifx/packets/stateTileState64.js:
```diff
@@ -0,0 +1,53 @@
+const Packet = {
+  size:
+    1 + 1 + 1 + 1 + 1 +
+    64 * (2 + 2 + 2 + 2),
```
The `size` is quite confusing like this. Just use the end result value like on the other packets.

mabels replied:
I simply changed it to the value.
To put such things in the comments is not helpful.
So we lose the source of this arbitrated value, i can live with this.

ristomatti commented on src/lifx/utils.js:
```diff
@@ -125,20 +186,22 @@ utils.rgbHexStringToObject = function(rgbHexString) {
-    return a - b;
-  });
-  return sortedCopy[0];
+  return Math.min.apply(undefined, array);
```
This is a nice cleanup but if the current babel settings allow, use spread instead of `apply`? It's not that important though.

mabels replied:
mist remerge with the on of the prev PR.
fixed

ristomatti reviewed (changes requested):
@mabels Added a couple of comments but I'll check the other changes after you've rebased the branch.

Sawtaytoes commented:
> Sorry a somewhat off-topic note:
> 
> > Also note that you can use arrow functions even though it's not used much if at all on the existing codebase (the ES2015 rewrite was done quite literally overnight). At least I would prefer arrow functions especially when passing an inline function as a parameter. I can't speak for @Sawtaytoes though who set up this fork though.
> 
> Quoting myself - now when I started to review the changes @mabels has done, I realize there's still two open PR's from me that's aren't included in this repo. When commenting about the arrow functions, I had an odd feeling like I would have done "something" to that already: [MariusRumpf/node-lifx#62](https://github.com/MariusRumpf/node-lifx/pull/62). This is the _rewrite_ I was talking about.
> 
> @Sawtaytoes It looks like I forgot PR 62 completely. Would it make sense for me to try to rebase it and do a PR? Or do you prefer working without the use of classes. I've personally been leaning towards more functinal programming approach (mainly at work) but I think using classes would make sense given how the library is written. There'd be a lot less clutter.

Whatever your preference is fine. I think bringing in those old PRs from your other repo would be fantastic!

I also like arrow functions. As long as our Node.js compatibility is fine, then we're good to go; otherwise, we should just bring up the min Node.js version and everything will be easier going forward.

mabels commented:
Node and Arrowfunctions: 

for i in boron carbon dubnium                                            ST 1   master
do
nvm run lts/$i -e '(() => console.log(process.versions.node))()'
done
Running node v6.17.1 (npm v3.10.10)
Running node v8.16.1 (npm v6.4.1)
Running node v10.16.3 (npm v6.11.3)

we should never go to 4.x so on my expirence we are fine.
We could also rewrite arrow functions with babel

mabels commented:
> > Sorry a somewhat off-topic note:
> > > Also note that you can use arrow functions even though it's not used much if at all on the existing codebase (the ES2015 rewrite was done quite literally overnight). At least I would prefer arrow functions especially when passing an inline function as a parameter. I can't speak for @Sawtaytoes though who set up this fork though.
> > 
> > 
> > Quoting myself - now when I started to review the changes @mabels has done, I realize there's still two open PR's from me that's aren't included in this repo. When commenting about the arrow functions, I had an odd feeling like I would have done "something" to that already: [MariusRumpf/node-lifx#62](https://github.com/MariusRumpf/node-lifx/pull/62). This is the _rewrite_ I was talking about.
> > @Sawtaytoes It looks like I forgot PR 62 completely. Would it make sense for me to try to rebase it and do a PR? Or do you prefer working without the use of classes. I've personally been leaning towards more functinal programming approach (mainly at work) but I think using classes would make sense given how the library is written. There'd be a lot less clutter.
> 
> Whatever your preference is fine. I think bringing in those old PRs from your other repo would be fantastic!
> 
> I also like arrow functions. As long as our Node.js compatibility is fine, then we're good to go; otherwise, we should just bring up the min Node.js version and everything will be easier going forward.

we should integrate "Arrow functions completly" after this PR is on master.
I offer my help.

ristomatti commented on example/tile.js:
```diff
@@ -0,0 +1,60 @@
+  console.log('New light found.');
+  console.log('ID: ' + light.id);
+
+  light.getDeviceChain(function(err, chain) {
```
This would now attempt to call `setTileState64` on any light. Add a `const TILE_LABEL = 'MyTile';` and trigger only on a light matching the label.

mabels replied:
done

ristomatti replied:
This does not work as light label has not resolved at this point. It was partly my bad as I didn't remember this while reviewing this the first time. It would improve the example a lot if a label could be used but for this purpose it would be enough to just rename `TILE_LABEL` TO `TILE_LIGHT_ID` for example.

ristomatti commented on example/tile.js:
```diff
@@ -0,0 +1,60 @@
+  });
+}
+
+function setBits(light, idx, chain) {
```
As this example has quite a lot going on, it would be great if you could add a comment describing what to expect when this starts running. Maybe include some `console.log`'s printing the values passed to `setTileState64` for example.

mabels replied:
done

ristomatti commented on src/lifx/light.js:
```diff
@@ -529,4 +529,162 @@ Light.prototype.colorZones = function(startIndex, endIndex, hue, saturation, bri
+  }
+  if (userY === undefined || typeof userY !== 'number') {
+    throw new TypeError('LIFX light setUserPosition userY to be a number');
+  }
```
These validations are now repeated on multiple functions and unnecessarily clutter the function body. Move these to new validation functions in `validator.js` similarly to what's been done for light zone support for example and then replace these with them.

mabels replied:
done

ristomatti commented on src/lifx/light.js:
```diff
@@ -529,4 +529,162 @@ Light.prototype.colorZones = function(startIndex, endIndex, hue, saturation, bri
+ * from the tileIndex. This will result in a separate response from
+ * each tile.
+ * For the LIFX Tile it really only makes sense to set x and y to
+ * zero, and width to 8.
```
If this comment holds, would it then make sense to move `x`, `y` and `width` as the last parameters and give them a default value using the `function(foo, bar = 'bar')` syntax? If the `babel` preset we have now does not support that, we ought to fix that first. If there's a specific reason for the current order of the parameters, please let me know.

I wouldn't want to rush this change as people might start creating pretty complex animations etc., after which it would be quite rude to change it.

mabels replied:
i change the api of set/get TileState64 to an option prop style
where the defaults are applied safely.

ristomatti reviewed (commented):
Added few comments. My JS limit for today has maxed out now though. I'll see if I could get this running tomorrow or Sunday.

Sawtaytoes commented:
@mabels @ristomatti Just your friendly neighborhood ping :).

ristomatti commented:
> @mabels @ristomatti Just your friendly neighborhood ping :).

I'm still waiting for @mabels. I guess eventually I could try sorting the issues myself but haven't had time to think about this too much unfortunately.

On a related note, I ran across this stuff that's based on the "competing" node-lifx-lan lib that looked pretty cool. Haven't tested it though:
- https://github.com/furey/lifx-tile-effects-framework
- https://github.com/furey/lifx-tile-halloween-effects

mabels committed:
changed api of set/getTileState64 to have an options
object and callback.
removed validation clutter from set/getTileState64
added some comments in the example

ristomatti commented:
Thanks @mabels. I will check and test this later today (= within the next 12 hours). If I won't, you're welcome to merge this @Sawtaytoes.

Sawtaytoes commented:
@ristomatti Merging in 1 hour.

ristomatti commented:
@Sawtaytoes Thanks for the heads-up (although I saw it only just now). Sadly I was blocked by a force majeure.

Would it be possible for me to email or PM you on this? You can find my email from my commit messages. Your location indicates you're 8 hours behind but I'm a night owl so there should be plenty of hours reach me.

ristomatti commented:
Impossible things can happen - testing now.

ristomatti commented:
@mabels Did you test this after your changes? I can't get this to work. There's multiple issues. When running `example/tile.js` I got this:

```
$ node tile.js 
Started LIFX listening on 0.0.0.0:49419

New light found.
ID: d073d5394495
/home/ristomatti/Projects/lifx-lan-client/lib/lifx/light.js:657
  validate.isUint8(tileIndex, 'setTileState64', 'tileIndex');
           ^

TypeError: validate.isUint8 is not a function
    at Light.setTileState64 (/home/ristomatti/Projects/lifx-lan-client/lib/lifx/light.js:657:12)
    at setBits (/home/ristomatti/Projects/lifx-lan-client/example/tile.js:41:9)
    at /home/ristomatti/Projects/lifx-lan-client/example/tile.js:61:5
    at Client.<anonymous> (/home/ristomatti/Projects/lifx-lan-client/lib/lifx/light.js:534:12)
```

After replacing renaming 3 calls to `validate.isUint8` with `validate.isUInt8` it crashed like so:

```
$ node tile.js 
Started LIFX listening on 0.0.0.0:43892

New light found.
ID: d073d5394495
/home/ristomatti/Projects/lifx-lan-client/lib/lifx/utils.js:143
    throw new TypeError('LIFX util buildColorsHsbk expects colors to be an array');
    ^

TypeError: LIFX util buildColorsHsbk expects colors to be an array
    at Object.utils.buildColorsHsbk (/home/ristomatti/Projects/lifx-lan-client/lib/lifx/utils.js:143:11)
    at Light.setTileState64 (/home/ristomatti/Projects/lifx-lan-client/lib/lifx/light.js:671:27)
    at setBits (/home/ristomatti/Projects/lifx-lan-client/example/tile.js:41:9)
    at /home/ristomatti/Projects/lifx-lan-client/example/tile.js:61:5
    at Client.<anonymous> (/home/ristomatti/Projects/lifx-lan-client/lib/lifx/light.js:534:12)
```

ristomatti commented:
@Sawtaytoes Sorry but this is not ready yet. I don't think we should rush this in as there might be other issues also. For instance `README.md` would need to be updated.

ristomatti commented on src/lifx/light.js:
```diff
@@ -529,4 +529,159 @@ Light.prototype.colorZones = function(startIndex, endIndex, hue, saturation, bri
+ * @param {Number} tileIndex	unsigned 8-bit integer
+ * @param {Number} userX	32-bit float
+ * @param {Number} userY	32-bit float
+ * @param {Number} reserved	16-bit integer
```
What is this `reserved` value and how would a user of this library know what to put there? This applies to all functions.

ristomatti commented on src/lifx/light.js:
```diff
@@ -529,4 +529,159 @@ Light.prototype.colorZones = function(startIndex, endIndex, hue, saturation, bri
+  if (typeof options === 'function') {
+    callback = options;
+    options = {};
+  }
```
What's the purpose of the above block?

mabels replied:
the options are optional  that you could pass on parameter options the callback

ristomatti commented on src/lifx/light.js:
```diff
@@ -529,4 +529,159 @@ Light.prototype.colorZones = function(startIndex, endIndex, hue, saturation, bri
+  if (typeof options === 'function') {
+    callback = options;
+    options = {};
+  }
```
What's the purpose of the above block? Also note that `options` is assigned to `{}` but `options` is not used after this.

mabels replied:
this was a typo

mabels commented:
with the unused options that is my fault.

I will fix it and build a test that it not happend again

meno

On Sat, 14 Dec 2019, 16:33 Ristomatti Airo, <notifications@github.com>
wrote:

> *@ristomatti* requested changes on this pull request.
> ------------------------------
>
> In src/lifx/light.js
> <https://github.com/node-lifx/lifx-lan-client/pull/15#discussion_r357923457>
> :
>
> > +  const sqnNumber = this.client.send(packetObj);
> +  this.client.addMessageHandler('stateDeviceChain', function(err, msg) {
> +    if (err) {
> +      return callback(err, null);
> +    }
> +    return callback(null, msg);
> +  }, sqnNumber);
> +};
> +
> +/**
> + * Sets Tile Position
> + * @example light.setUserPosition(0, 12, 13, 0, () => {})
> + * @param {Number} tileIndex	unsigned 8-bit integer
> + * @param {Number} userX	32-bit float
> + * @param {Number} userY	32-bit float
> + * @param {Number} reserved	16-bit integer
>
> What is this reserved value and how would a user of this library know
> what to put there? This applies to all functions.
> ------------------------------
>
> In src/lifx/light.js
> <https://github.com/node-lifx/lifx-lan-client/pull/15#discussion_r357923685>
> :
>
> > + * each tile.
> + * @param {Number} tileIndex	unsigned 8-bit integer
> + * @param {Object} options - options passed to
> + * @param {Number} options.length	unsigned 8-bit integer (default 64)
> + * @param {Number} options.x	unsigned 8-bit integer (default 0)
> + * @param {Number} options.y	unsigned 8-bit integer (default 0)
> + * @param {Number} options.width	unsigned 8-bit integer (default 8)
> + * @param {Number} options.reserved	unsigned 8-bit integer
> + * @param {Function} callback a function to accept the data
> + */
> +Light.prototype.getTileState64 = function(tileIndex, options, callback) {
> +  validate.isUint8(tileIndex, 'getTileState64', 'tileIndex');
> +  if (typeof options === 'function') {
> +    callback = options;
> +    options = {};
> +  }
>
> What's the purpose of the above block?
> ------------------------------
>
> In src/lifx/light.js
> <https://github.com/node-lifx/lifx-lan-client/pull/15#discussion_r357923757>
> :
>
> > + * @param {Number} colors[64]	64 HSBK values
> + * @param {Object} options - options passed to
> + * @param {Number} options.length	unsigned 8-bit integer (default 0)
> + * @param {Number} options.x	unsigned 8-bit integer (default 8)
> + * @param {Number} options.y	unsigned 8-bit integer (default 8)
> + * @param {Number} options.width	unsigned 8-bit integer (default 8)
> + * @param {Number} options.duration	unsigned 32-bit integer (default 0)
> + * @param {Number} options.reserved	unsigned 8-bit integer
> + * @param {Function} [callback] called when light did receive message
> + */
> +Light.prototype.setTileState64 = function(tileIndex, colors, options, callback) {
> +  validate.isUint8(tileIndex, 'setTileState64', 'tileIndex');
> +  if (typeof options === 'function') {
> +    callback = options;
> +    options = {};
> +  }
>
> What's the purpose of the above block? Also note that options is assigned
> to {} but options is not used after this.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/node-lifx/lifx-lan-client/pull/15?email_source=notifications&email_token=AAAEWJWF5L7P4JR5RRZJP3TQYT4DJA5CNFSM4IWUA6Y2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCPGUDFY#pullrequestreview-332218775>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAEWJRMQTAPM6KLROLHV2DQYT4DJANCNFSM4IWUA6YQ>
> .
>

mabels committed:
fix in setTileState64 the missing option passing

mabels commented:
> @mabels Did you test this after your changes? I can't get this to work. There's multiple issues. When running `example/tile.js` I got this:
> 
> ```
> $ node tile.js 
> Started LIFX listening on 0.0.0.0:49419
> 
> New light found.
> ID: d073d5394495
> /home/ristomatti/Projects/lifx-lan-client/lib/lifx/light.js:657
>   validate.isUint8(tileIndex, 'setTileState64', 'tileIndex');
>            ^
> 
> TypeError: validate.isUint8 is not a function
>     at Light.setTileState64 (/home/ristomatti/Projects/lifx-lan-client/lib/lifx/light.js:657:12)
>     at setBits (/home/ristomatti/Projects/lifx-lan-client/example/tile.js:41:9)
>     at /home/ristomatti/Projects/lifx-lan-client/example/tile.js:61:5
>     at Client.<anonymous> (/home/ristomatti/Projects/lifx-lan-client/lib/lifx/light.js:534:12)
> ```
> 
> After replacing renaming 3 calls to `validate.isUint8` with `validate.isUInt8` it crashed like so:
> 
> ```
> $ node tile.js 
> Started LIFX listening on 0.0.0.0:43892
> 
> New light found.
> ID: d073d5394495
> /home/ristomatti/Projects/lifx-lan-client/lib/lifx/utils.js:143
>     throw new TypeError('LIFX util buildColorsHsbk expects colors to be an array');
>     ^
> 
> TypeError: LIFX util buildColorsHsbk expects colors to be an array
>     at Object.utils.buildColorsHsbk (/home/ristomatti/Projects/lifx-lan-client/lib/lifx/utils.js:143:11)
>     at Light.setTileState64 (/home/ristomatti/Projects/lifx-lan-client/lib/lifx/light.js:671:27)
>     at setBits (/home/ristomatti/Projects/lifx-lan-client/example/tile.js:41:9)
>     at /home/ristomatti/Projects/lifx-lan-client/example/tile.js:61:5
>     at Client.<anonymous> (/home/ristomatti/Projects/lifx-lan-client/lib/lifx/light.js:534:12)
> ```

I could not test this change off course i had no lifx on hand.
I will try this if i come back home in the next week.

mabels committed:
fix: the typo of Uint8 to UInt8

mabels commented:
> > @mabels Did you test this after your changes? I can't get this to work. There's multiple issues. When running `example/tile.js` I got this:
> > ```
> > $ node tile.js 
> > Started LIFX listening on 0.0.0.0:49419
> > 
> > New light found.
> > ID: d073d5394495
> > /home/ristomatti/Projects/lifx-lan-client/lib/lifx/light.js:657
> >   validate.isUint8(tileIndex, 'setTileState64', 'tileIndex');
> >            ^
> > 
> > TypeError: validate.isUint8 is not a function
> >     at Light.setTileState64 (/home/ristomatti/Projects/lifx-lan-client/lib/lifx/light.js:657:12)
> >     at setBits (/home/ristomatti/Projects/lifx-lan-client/example/tile.js:41:9)
> >     at /home/ristomatti/Projects/lifx-lan-client/example/tile.js:61:5
> >     at Client.<anonymous> (/home/ristomatti/Projects/lifx-lan-client/lib/lifx/light.js:534:12)
> > ```
> > 
> > 
> > After replacing renaming 3 calls to `validate.isUint8` with `validate.isUInt8` it crashed like so:
> > ```
> > $ node tile.js 
> > Started LIFX listening on 0.0.0.0:43892
> > 
> > New light found.
> > ID: d073d5394495
> > /home/ristomatti/Projects/lifx-lan-client/lib/lifx/utils.js:143
> >     throw new TypeError('LIFX util buildColorsHsbk expects colors to be an array');
> >     ^
> > 
> > TypeError: LIFX util buildColorsHsbk expects colors to be an array
> >     at Object.utils.buildColorsHsbk (/home/ristomatti/Projects/lifx-lan-client/lib/lifx/utils.js:143:11)
> >     at Light.setTileState64 (/home/ristomatti/Projects/lifx-lan-client/lib/lifx/light.js:671:27)
> >     at setBits (/home/ristomatti/Projects/lifx-lan-client/example/tile.js:41:9)
> >     at /home/ristomatti/Projects/lifx-lan-client/example/tile.js:61:5
> >     at Client.<anonymous> (/home/ristomatti/Projects/lifx-lan-client/lib/lifx/light.js:534:12)
> > ```
> 
> I could not test this change off course i had no lifx on hand.
> I will try this if i come back home in the next week.

mabels commented:
I need to build a test for the high order function. I'm sorry for this chain of failures.
